statistics/statistics.go

```diff
@@ -212,94 +293,133 @@ func (c *Column) mergeBuckets(bucketIdx int64) {
-		Ndv:     proto.Int64(col.NDV),
-		Numbers: col.Numbers,
-		Value:   data,
-		Repeats: col.Repeats,
+	insertSQL := fmt.Sprintf("insert into mysql.stats_meta (version, table_id, count) values (%d, %d, %d)", version, t.Info.ID, t.Count)
+	_, err = ctx.(sqlexec.SQLExecutor).Execute(insertSQL)
+	if err != nil {
+		return errors.Trace(err)
 	}
-	return cpb, nil
-}
-
-// columnFromPB gets Column from ColumnPB.
-func columnFromPB(cpb *ColumnPB, ft *types.FieldType) (*Column, error) {
-	var values []types.Datum
-	var err error
-	if len(cpb.GetValue()) > 0 {
-		values, err = codec.Decode(cpb.GetValue(), 1)
+	deleteSQL = fmt.Sprintf("delete from mysql.stats_columns where table_id = %d", t.Info.ID)
+	_, err = ctx.(sqlexec.SQLExecutor).Execute(deleteSQL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	deleteSQL = fmt.Sprintf("delete from mysql.stats_buckets where table_id = %d", t.Info.ID)
+	_, err = ctx.(sqlexec.SQLExecutor).Execute(deleteSQL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, col := range t.Columns {
+		err = col.saveToStorage(ctx, t.Info.ID, false)
 		if err != nil {
-			return nil, errors.Trace(err)
+			return errors.Trace(err)
 		}
 	}
-	c := &Column{
-		ID:      cpb.GetId(),
-		NDV:     cpb.GetNdv(),
-		Numbers: cpb.GetNumbers(),
-		Values:  make([]types.Datum, len(values)),
-		Repeats: cpb.GetRepeats(),
-	}
-	for i, val := range values {
-		c.Values[i], err = tablecodec.Unflatten(val, ft, false)
+	for _, idx := range t.Indices {
+		err = idx.saveToStorage(ctx, t.Info.ID, true)
 		if err != nil {
-			return nil, errors.Trace(err)
+			return errors.Trace(err)
 		}
 	}
-	return c, nil
+	_, err = ctx.(sqlexec.SQLExecutor).Execute("commit")
+	return errors.Trace(err)
 }
 
-// Table represents statistics for a table.
-type Table struct {
-	Info    *model.TableInfo
-	TS      int64 // build timestamp.
-	Columns []*Column
-	Indices []*Column
-	Count   int64 // Total row count in a table.
-	Pseudo  bool
+// TableStatsFromStorage load table stats info from storage.
+func TableStatsFromStorage(ctx context.Context, info *model.TableInfo, count int64) (*Table, error) {
+	table := &Table{
+		Info:  info,
+		Count: count,
+	}
+	selSQL := fmt.Sprintf("select * from mysql.stats_columns where table_id = %d", info.ID)
+	rows, _, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, selSQL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// indexCount and columnCount records the number of indices and columns in table stats. If the number don't match with
+	// tableInfo, we will return pseudo table.
+	// TODO: In fact, we can return pseudo column.
+	indexCount, columnCount := 0, 0
+	for _, row := range rows {
+		distinct := row.Data[3].GetInt64()
+		if row.Data[1].IsNull() {
+			// process index
+			idxID := row.Data[2].GetInt64()
+			var col *Column
+			for _, idxInfo := range info.Indices {
+				if idxID == idxInfo.ID {
+					col, err = colStatsFromStorage(ctx, info.ID, idxID, nil, distinct, true)
```

Comment:
If line [120](https://github.com/pingcap/tidb/blob/5effd89cafb083452e899861665821c45789e1f5/statistics/statistics.go#L120) return nil, this means that this index do not have histograms as line [727](https://github.com/pingcap/tidb/blob/5effd89cafb083452e899861665821c45789e1f5/statistics/statistics.go#L727) shows, as those who have histograms must have at least 1 bucket even if it is empty columns. So I suggest that line [120](https://github.com/pingcap/tidb/blob/5effd89cafb083452e899861665821c45789e1f5/statistics/statistics.go#L120) return the psedo cols, and update the comment in [727](https://github.com/pingcap/tidb/blob/5effd89cafb083452e899861665821c45789e1f5/statistics/statistics.go#L727).